src/main/java/org/opensearch/flowframework/util/WorkflowTimeoutUtility.java

```diff
@@ -0,0 +1,198 @@
+        }
+    }
+
+    /**
+     * Wraps a listener with a timeout cancellation listener to cancel the timeout task when the workflow completes.
+     *
+     * @param listener             The original listener to wrap.
+     * @param scheduledCancellable The cancellable timeout task.
+     * @param isResponseSent        An atomic boolean to ensure the response is sent only once.
+     * @param <Response>            The type of the response expected by the listener.
+     * @return A wrapped ActionListener with timeout cancellation logic.
+     */
+    public static <Response> ActionListener<Response> wrapWithTimeoutCancellationListener(
+        ActionListener<Response> listener,
+        Scheduler.ScheduledCancellable scheduledCancellable,
+        AtomicBoolean isResponseSent
+    ) {
+        return new ActionListener<>() {
+            @Override
+            public void onResponse(Response response) {
+                if (isResponseSent.compareAndSet(false, true)) {
+                    scheduledCancellable.cancel();
+                }
+                listener.onResponse(response);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                if (isResponseSent.compareAndSet(false, true)) {
+                    scheduledCancellable.cancel();
+                }
+                listener.onFailure(e);
+            }
+        };
+    }
+
+    /**
+     * Handles the successful completion of a workflow.
+     *
+     * @param workflowId     The unique identifier of the workflow.
+     * @param workflowResponse The response from the workflow execution.
+     * @param isResponseSent  An atomic boolean to ensure the response is sent only once.
+     * @param listener        The listener to notify of the workflow completion.
+     */
+    public static void handleResponse(
+        String workflowId,
+        WorkflowResponse workflowResponse,
+        AtomicBoolean isResponseSent,
+        ActionListener<WorkflowResponse> listener
+    ) {
+        if (isResponseSent.compareAndSet(false, true)) {
+            listener.onResponse(new WorkflowResponse(workflowResponse.getWorkflowId(), workflowResponse.getWorkflowState()));
+        } else {
+            logger.info("Ignoring onResponse for workflowId: {} as timeout already occurred", workflowId);
+        }
+    }
+
+    /**
+     * Handles the failure of a workflow execution.
+     *
+     * @param workflowId    The unique identifier of the workflow.
+     * @param e             The exception that occurred during workflow execution.
+     * @param isResponseSent An atomic boolean to ensure the response is sent only once.
+     * @param listener       The listener to notify of the workflow failure.
+     */
+    public static void handleFailure(
+        String workflowId,
+        Exception e,
+        AtomicBoolean isResponseSent,
+        ActionListener<WorkflowResponse> listener
+    ) {
+        if (isResponseSent.compareAndSet(false, true)) {
```

Comment:
For those of us who don't use AtomicBooleans often, a comment above this saying what we're doing would be helpful.   (I think I can guess from the else clause.)